### ISSUE TITLE
Have default value for fromBase Base type

### ISSUE THREAD

**Issue by JoepKockelkorn** [anton-kravchenko/ObjectBuilder#20]
Hi, thanks for your great utility library!

We've noticed that we are constantly giving the same generic arguments to the `fromBase` method, i.e. `fromBase<MyType, MyType>` as we don't have partial fixtures, only full ones. It would be nice if the second parameter 'Base' would have the default type of the type guard: `Partial<Target>` so we don't have to specify it twice redundantly everywhere.

So proposed solution: `fromBase<Target extends object, Base extends Partial<Target> = Partial<Target>>`.
Reason: improved ease of use by having sensible defaults.

I can make a PR if you like?

**Comment by anton-kravchenko** [comment]
Hi @JoepKockelkorn. I'm sorry for the late response. Thank you for using the library and suggesting a way to improve it.

I've tried to implement the suggested change, and unfortunately, it didn't work for me. Please see the code below:

```
// with `public static fromBase<Target extends object, Base extends Partial<Target> = Partial<Target>>(`

type TestType = { method: string; status: number; response: string };
const base = { method: 'get' };
ObjectBuilder.fromBase<TestType>(base).build(); // doesn't throw error although it should since `base` is missing 2 fields
```
I would appreciate it if you could implement the suggested change (in case I missed something). The "test cases" are located at `src/tests/type-check-test/type-check-object-builder-fromBase.ts` (please make sure that all of the test cases remain working after the introduction of new changes).

Meanwhile, I've implemented a new `basedOn` method that does what you requested (`v0.1.1`). See the example below:
```
type T = { status: number; response: string };
const base: T = { status: 200, response: '' };

const foo: T = ObjectBuilder.basedOn(base).build();
const bar: T = ObjectBuilder.basedOn(base).with('response', '{}').build();
const baz: T = ObjectBuilder.basedOn<T>(base).with('response', '{}').with('status', 201).build();
```

The type argument is optional, and the API lets you overwrite any fields you need and then build your object. 
This API is added to `v0.1.1`. I haven't documented this new method yet - I wanted to get your feedback first.

**Comment by JoepKockelkorn** [comment]
Hi @anton-kravchenko, thanks for you response. I'll try if I can make it work in the existing `fromBase` method, but I'm quite busy so I don't know when I can find the time. Thanks for the effort on the new method! If you want you can close this issue.

**Comment by JoepKockelkorn** [comment]
@anton-kravchenko My assumption was wrong, it should not be a `Partial<Target>`, but a full one(`Target`). I've made [a draft PR](https://github.com/anton-kravchenko/ObjectBuilder/pull/23) with a potential solution. If this passes all your checks it might also even be possible to reduce some typings by using default generic args.

**Comment by anton-kravchenko** [comment]
@JoepKockelkorn, thank you for the PR. It works as expected and passes all checks. The change has been released to the npm registry as `v0.3.0`.

Feel free to share any ideas regarding improving/optimizing/reducing type declarations - the implementation is really tricky, and any optimizations are welcome :) 

Also, if you have any ideas regarding extending the API - feel free to share.

**Comment by JoepKockelkorn** [comment]
@anton-kravchenko No problem, will do if I have time. I'll close this issue as it's been fixed!
